In [1]:
# Initializing spark session

spark

VBox()

Starting Spark application


FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

SparkSession available as 'spark'.


FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [2]:
# Importing required libraries

from pyspark.sql.functions import *
from pyspark.sql import Window
from pyspark.sql.types import StructType, StructField, IntegerType, StringType, BooleanType, DoubleType, LongType

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [3]:
# Setting up the file schema to import raw data into a dataframe

fileSchema = StructType([StructField('year',IntegerType(),True),
                        StructField('month',StringType(),True),
                        StructField("day",IntegerType(),True),
                        StructField("weekday",StringType(),True),
                        StructField("hour",IntegerType(),True),
                        StructField("atm_status",StringType(),True),
                        StructField("atm_id",StringType(),True),
                        StructField("atm_manufacturer",StringType(),True),
                        StructField("atm_location",StringType(),True),
                        StructField("atm_streetname",StringType(),True),
                        StructField("atm_street_number",IntegerType(),True),
                        StructField("atm_zipcode",IntegerType(),True),
                        StructField("atm_lat",DoubleType(),True),
                        StructField("atm_lon",DoubleType(),True),
                        StructField("currency",StringType(),True),
                        StructField("card_type",StringType(),True),
                        StructField("transaction_amount",IntegerType(),True),
                        StructField("service",StringType(),True),
                        StructField("message_code",StringType(),True),
                        StructField("message_text",StringType(),True),
                        StructField("weather_lat",DoubleType(),True),
                        StructField("weather_lon",DoubleType(),True),
                        StructField("weather_city_id",IntegerType(),True),
                        StructField("weather_city_name",StringType(),True),
                        StructField("temp",DoubleType(),True),
                        StructField("pressure",IntegerType(),True),
                        StructField("humidity",IntegerType(),True),
                        StructField("wind_speed",IntegerType(),True),
                        StructField("wind_deg",IntegerType(),True),
                        StructField("rain_3h",DoubleType(),True),
                        StructField("clouds_all",IntegerType(),True),
                        StructField("weather_id",IntegerType(),True),
                        StructField("weather_main",StringType(),True),
                        StructField("weather_description",StringType(),True)])

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [4]:
# Importing data into a dataframe from HDFS

atm_data = spark.read.csv("hdfs:/user/hadoop/etl_project", header = False , schema = fileSchema)

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [5]:
# Checking number of records in the dataframe

atm_data.count()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

2468572

In [6]:
# Looking at the schema of the dataframe

atm_data.printSchema()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

root
 |-- year: integer (nullable = true)
 |-- month: string (nullable = true)
 |-- day: integer (nullable = true)
 |-- weekday: string (nullable = true)
 |-- hour: integer (nullable = true)
 |-- atm_status: string (nullable = true)
 |-- atm_id: string (nullable = true)
 |-- atm_manufacturer: string (nullable = true)
 |-- atm_location: string (nullable = true)
 |-- atm_streetname: string (nullable = true)
 |-- atm_street_number: integer (nullable = true)
 |-- atm_zipcode: integer (nullable = true)
 |-- atm_lat: double (nullable = true)
 |-- atm_lon: double (nullable = true)
 |-- currency: string (nullable = true)
 |-- card_type: string (nullable = true)
 |-- transaction_amount: integer (nullable = true)
 |-- service: string (nullable = true)
 |-- message_code: string (nullable = true)
 |-- message_text: string (nullable = true)
 |-- weather_lat: double (nullable = true)
 |-- weather_lon: double (nullable = true)
 |-- weather_city_id: integer (nullable = true)
 |-- weather_city_name: st

In [7]:
# Inspecting the first 5 rows of the dataframe

atm_data.show(n=5)

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

+----+-------+---+-------+----+----------+------+----------------+------------+-------------------+-----------------+-----------+-------+-------+--------+----------+------------------+----------+------------+------------+-----------+-----------+---------------+-----------------+------+--------+--------+----------+--------+-------+----------+----------+------------+--------------------+
|year|  month|day|weekday|hour|atm_status|atm_id|atm_manufacturer|atm_location|     atm_streetname|atm_street_number|atm_zipcode|atm_lat|atm_lon|currency| card_type|transaction_amount|   service|message_code|message_text|weather_lat|weather_lon|weather_city_id|weather_city_name|  temp|pressure|humidity|wind_speed|wind_deg|rain_3h|clouds_all|weather_id|weather_main| weather_description|
+----+-------+---+-------+----+----------+------+----------------+------------+-------------------+-----------------+-----------+-------+-------+--------+----------+------------------+----------+------------+------------+-

### Creating Dimension Tables

##### 1. Location Dimesion Table

In [8]:
# Selecting required columns for location dimension table and deleting duplicate records

LOCATION_DIM = atm_data.select('atm_location','atm_streetname','atm_street_number','atm_zipcode','atm_lat','atm_lon')
LOCATION_DIM = LOCATION_DIM.dropDuplicates((['atm_location','atm_streetname','atm_street_number', 'atm_zipcode', 'atm_lat', 'atm_lon']))

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [9]:
# Creating a primary key 'location_id' for the location dimension table

LOCATION_DIM = LOCATION_DIM.withColumn('location_id',row_number().over(Window.orderBy(monotonically_increasing_id())))

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [10]:
# Re-arraging column order

LOCATION_DIM = LOCATION_DIM.select('location_id','atm_location','atm_streetname','atm_street_number','atm_zipcode','atm_lat','atm_lon')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [11]:
# Checking the schema of location dimension table

LOCATION_DIM.printSchema()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

root
 |-- location_id: integer (nullable = true)
 |-- atm_location: string (nullable = true)
 |-- atm_streetname: string (nullable = true)
 |-- atm_street_number: integer (nullable = true)
 |-- atm_zipcode: integer (nullable = true)
 |-- atm_lat: double (nullable = true)
 |-- atm_lon: double (nullable = true)

In [12]:
# Checking the counts of location dimension table

LOCATION_DIM.count()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

109

##### 2. Creating DATE Dimention Table

In [13]:
# Selecting only the required columns dor date dimension table

DATE_DIM = atm_data.select('year','month','day','hour','weekday')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [14]:
# Creating full_date_column by concating the month, day, year and timsestamp for uniquely identifying each record

DATE_DIM = DATE_DIM.withColumn("month_numeric", from_unixtime(unix_timestamp(col("month"),'MMM'),'MM'))
DATE_DIM = DATE_DIM.withColumn('full_date_time',concat(col('year'),lit('/'),col('month_numeric'),lit('/'),col('day'),lit(' '),col('hour'),lit(':'),lit('00:00')))
DATE_DIM = DATE_DIM.withColumn('full_date_time', unix_timestamp(DATE_DIM['full_date_time'], 'yyyy/MM/dd HH:mm:ss').cast('timestamp'))

# Dropping the month_numeric column as it is not reuqired further

DATE_DIM = DATE_DIM.drop('month_numeric')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [15]:
# Dropping duplicate records based on full_date_time

DATE_DIM = DATE_DIM.dropDuplicates((['full_date_time']))

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [16]:
# Adding a primary key 'date_id' for date dimension table

DATE_DIM = DATE_DIM.withColumn(
    "date_id",
    row_number().over(Window.orderBy(monotonically_increasing_id()))
)

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [17]:
# Looking at the schema for DATE dimension table

DATE_DIM.printSchema()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

root
 |-- year: integer (nullable = true)
 |-- month: string (nullable = true)
 |-- day: integer (nullable = true)
 |-- hour: integer (nullable = true)
 |-- weekday: string (nullable = true)
 |-- full_date_time: timestamp (nullable = true)
 |-- date_id: integer (nullable = true)

In [18]:
# Checking count for date dimension

DATE_DIM.count()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

8685

##### 3. Creating ATM Dimension Table

In [19]:
# Selecting required columns for ATM dimension and dropping duplicate records

ATM_DIM = atm_data.select('atm_id','atm_manufacturer','atm_location','atm_streetname','atm_street_number')
ATM_DIM = ATM_DIM.dropDuplicates()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [20]:
# Joining ATM dimesntion with LOCATION dimension to get location_id column

ATM_DIM = ATM_DIM.join(LOCATION_DIM, on=['atm_location','atm_streetname','atm_street_number'],how='left')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [21]:
# Renaming atm_id to atm_number

ATM_DIM = ATM_DIM.withColumnRenamed('atm_id','atm_number')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [22]:
# Adding a primary key atm_id to ATM dimension table

ATM_DIM =  ATM_DIM.withColumn(
    "atm_id",
    row_number().over(Window.orderBy(monotonically_increasing_id()))
)


# Re-arraging column order

ATM_DIM = ATM_DIM.select('atm_id','atm_number','atm_manufacturer','location_id')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [23]:
ATM_DIM.printSchema()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

root
 |-- atm_id: integer (nullable = true)
 |-- atm_number: string (nullable = true)
 |-- atm_manufacturer: string (nullable = true)
 |-- location_id: integer (nullable = true)

In [24]:
# Checking the schema for ATM dimension

ATM_DIM.count()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

113

##### 4. Creating CARD_TYPE Dimension

In [25]:
# Selecting card_type columns for credit card dimension table and dropping duplicates

CREDIT_DIM = atm_data.select('card_type')
CREDIT_DIM = CREDIT_DIM.dropDuplicates()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [26]:
# Adding the primary key 'card_type_id' for credit card dimension table

CREDIT_DIM = CREDIT_DIM.withColumn('card_type_id',row_number().over(Window.orderBy(monotonically_increasing_id())))

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [27]:
# Checking the schema for credit card dimension table

CREDIT_DIM.printSchema()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

root
 |-- card_type: string (nullable = true)
 |-- card_type_id: integer (nullable = true)

In [48]:
# Checking the count for credit card dimension table

CREDIT_DIM.count()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

12

##### 5. Creating TRANSACTION Fact Table

In [29]:
# Selecting all the columns initially, will filter out as required

tempfact = atm_data.select("*")

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [30]:
# Joining with date dimension to get date_id

tempfact = tempfact.join(DATE_DIM, on=['year','month','day','hour','weekday'], how='left')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [31]:
# Joining with credit card dimension to get card_type_id

tempfact = tempfact.join(CREDIT_DIM, on = 'card_type', how = 'left')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [32]:
# Joining with location dimension to get location_id

tempfact = tempfact.join(LOCATION_DIM, on = ['atm_location', 'atm_lat', 'atm_lon', 'atm_streetname', 'atm_street_number', 
                                               'atm_zipcode'], how = 'left')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [33]:
# Renaming atm_id to atm_number and joining with atm dimension to get atm_id

tempfact = tempfact.withColumnRenamed('atm_id','atm_number')
tempfact = tempfact.join(ATM_DIM, on = ['atm_number', 'atm_manufacturer', 'location_id'], how = 'left')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [34]:
# Addind a primary key "trans_id" to the fact table for transactions

tempfact = tempfact.withColumn(
    'trans_id',
    row_number().over(Window.orderBy(monotonically_increasing_id())))

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [35]:
# selecting the required columns for transaction table nad inspecting the schema

TRANSACTION_FACT = tempfact.select('trans_id','atm_id','location_id','date_id','card_type_id','atm_status','currency','service',
                       'transaction_amount','message_code','message_text','rain_3h','clouds_all','weather_id','weather_main','weather_description')
TRANSACTION_FACT.printSchema()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

root
 |-- trans_id: integer (nullable = true)
 |-- atm_id: integer (nullable = true)
 |-- location_id: integer (nullable = true)
 |-- date_id: integer (nullable = true)
 |-- card_type_id: integer (nullable = true)
 |-- atm_status: string (nullable = true)
 |-- currency: string (nullable = true)
 |-- service: string (nullable = true)
 |-- transaction_amount: integer (nullable = true)
 |-- message_code: string (nullable = true)
 |-- message_text: string (nullable = true)
 |-- rain_3h: double (nullable = true)
 |-- clouds_all: integer (nullable = true)
 |-- weather_id: integer (nullable = true)
 |-- weather_main: string (nullable = true)
 |-- weather_description: string (nullable = true)

In [36]:
# Checking the counts for Fact Table

TRANSACTION_FACT.count()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

2468572

#### Making sure the counts match with the validation document

In [37]:
# Checking the count

TRANSACTION_FACT.count()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

2468572

In [38]:
# Checking the count

LOCATION_DIM.count()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

109

In [39]:
# Checking the count

ATM_DIM.count()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

113

In [40]:
# Checking the count

DATE_DIM.count()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

8685

In [41]:
# Checking the count

CREDIT_DIM.count()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

12

##### Summary of Counts:
- ATM_DIM : 113
- LOCATION_DIM : 109
- DATE_DIM : 8685
- CREDIT_DIM : 12
- TRANSACTION_FACT : 2468572

## Saving to S3

In [42]:
# Setting up access key and secret key for AWS s3 bucket

sc._jsc.hadoopConfiguration().set("fs.s3a.access.key","AKIAUG5COSBKBAWD7UUX")
sc._jsc.hadoopConfiguration().set("fs.s3a.secret.key","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")

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [43]:
# Writing location dimension table to s3

LOCATION_DIM.write.save("s3a://upgrad-etl-project/DIM_LOC",format='csv',header='true')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [44]:
# Writing atm dimension table to s3

ATM_DIM.write.save("s3a://upgrad-etl-project/DIM_ATM",format='csv',header='true')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [45]:
# Writing date dimension table to s3

DATE_DIM.write.save("s3a://upgrad-etl-project/DIM_DATE",format='csv',header='true')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [46]:
# Writing credit card dimension table to s3

CREDIT_DIM.write.save("s3a://upgrad-etl-project/DIM_CARD",format='csv',header='true')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [47]:
# Writing transaction fact table to s3

TRANSACTION_FACT.write.save("s3a://upgrad-etl-project/FACT_ATM_TRANS",format='csv',header='true')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…